uniswap-v2/logics/impls/router/router.rs

```diff
@@ -0,0 +1,505 @@
+        amount_b_min: Balance,
+        to: AccountId,
+        deadline: u64,
+    ) ->  Result<(Balance, Balance, Balance), RouterError> {
+        if deadline <= Self::env().block_timestamp() {
+            return Err(RouterError::Expired)
+        }
+
+        let (amount_a, amount_b) = self._add_liquidity(
+            token_a,
+            token_b,
+            amount_a_desired,
+            amount_b_desired,
+            amount_a_min,
+            amount_b_min,
+        )?;
+
+        let factory = self.data::<data::Data>().factory;
+        let pair_contract = self._pair_for(factory, token_a, token_b)?;
+
+        PSP22Ref::transfer_from(
+            &token_a,
+            Self::env().caller(),
+            pair_contract,
+            amount_a,
+            Vec::new(),
+        )?;
+        PSP22Ref::transfer_from(
+            &token_b,
+            Self::env().caller(),
+            pair_contract,
+            amount_b,
+            Vec::new(),
+        )?;
+
+        let liquidity = PairRef::mint(&pair_contract, to)?;
+        
+        ink_env::debug_println!("Liquidity {:?}", liquidity);
+
+        Ok((amount_a, amount_b, liquidity))
+    }
+
+    default fn remove_lequidity(
+        &mut self,
+        token_a: AccountId,
+        token_b: AccountId,
+        liquidity: Balance,
+        amount_a_min: Balance,
+        amount_b_min: Balance,
+        to: AccountId,
+        deadline: u64,
+    ) -> Result<(Balance, Balance), RouterError> {
+        if deadline <= Self::env().block_timestamp() {
+            return Err(RouterError::Expired)
+        }
+
+        let factory = self.data::<data::Data>().factory;
+        let pair_contract = self._pair_for(factory, token_a, token_b)?;
+
+        PSP22Ref::transfer_from(
+            &pair_contract,
+            Self::env().caller(),
+            pair_contract,
+            liquidity,
+            Vec::new(),
+        )?;
+
+        let (amount_0, amount_1) = PairRef::burn(&pair_contract, to)?;
+        let (token_0, _) = self._sort_tokens(token_a, token_b)?;
+        let (amount_a, amount_b) = if token_a == token_0 {
+            (amount_0, amount_1)
+        } else {
+            (amount_1, amount_0)
+        };
+
+        if amount_a < amount_a_min {
+            return Err(RouterError::InsufficientAAmount)
+        }
+        if amount_b < amount_b_min {
+            return Err(RouterError::InsufficientBAmount)
+        }
+
+        Ok((amount_a, amount_b))
+    }
+
+    default fn swap_exact_tokens_for_tokens(
+        &mut self,
+        amount_in: Balance,
+        amount_out_min: Balance,
+        path: Vec<AccountId>,
+        to: AccountId,
+        deadline: u64,
+    ) -> Result<Vec<Balance>, RouterError> {
+        if deadline <= Self::env().block_timestamp() {
+            return Err(RouterError::Expired)
+        }
+
+        let factory = self.data::<data::Data>().factory;
+        let amounts: Vec<Balance> = self._get_amounts_out(factory, amount_in, path.clone())?;
```

Comment:
no need to clone() here (cloning a Vec is expensive)
just refactor _get_amounts_out & _get_amounts_in to take a &Vec<AccountId> and pass a ref here (and wherever it is needed)